Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm

```diff
@@ -249,7 +249,34 @@ static bool experimentalFeatureEnabled(const String& key, bool defaultValue = fa
         websiteDataStoreDirectory.get() = [webkitDirectory URLByAppendingPathComponent:@"WebsiteDataStore" isDirectory:YES];
     });
 
-    return [websiteDataStoreDirectory.get() URLByAppendingPathComponent:identifier isDirectory:YES].absoluteURL.path;
+    return websiteDataStoreDirectory.get().get().absoluteURL.path;
+}
+
+void WebsiteDataStore::fetchAllDataStoreIdentifiers(CompletionHandler<void(Vector<UUID>&&)>&& completionHandler)
+{
+    Vector<UUID> identifiers;
+    for (auto identifierString : FileSystem::listDirectory(defaultWebsiteDataStoreRootDirectory())) {
+        if (auto identifier = UUID::parse(identifierString))
+            identifiers.append(*identifier);
+    }
+
+    completionHandler(WTFMove(identifiers));
+}
+
+void WebsiteDataStore::removeDataStoreWithIdentifier(const UUID& identifier, CompletionHandler<void(const String&)>&& completionHandler)
```

Comment:
Not for this patch, but: I wonder if we should do anything to post a message to the Network process and wait for an async reply, to make sure that any pending storage operations have been flushed before we attempt to remove the directory. The most basic solution would just post to the main thread of the Network process, to make sure to clear out any pending IPC. Even better would be to post a task to each storage type thread and wait for that set of tasks.

Reply:
> I think a simple test where you just set a cookie and then delete a data store may fail in the current implementation
Do you mean the directory might be failed to delete?

I guess it makes sense to ask child processes to clear data (or finish existing operations) before UI process deletes the directory when `WebsiteDataStore` with target identifier is alive. But even network process finishes task scheduled before `removeDataStoreWithIdentifier`, there seems no guarantee that network process will no longer write to that WebsiteDataStore's directories (e.g. ServiceWorker process updates LocalStorage).
Do you think we should forbid network process running storage operations on the session (WebsiteDataStore) after network process replies okay to `removeDataStoreWithIdentifier` message? We might achieve that by revoking sandbox extension to paths I guess.